Cookbook/Proposals/PropertyWrappers.md

```diff
@@ -0,0 +1,155 @@
+ 
+### Detailed design
+
+Swift compiler is capable of generating decoding code for properties wrapped with property wrappers in a non disruptive way (i.e. it does not require changing coding keys to match names of wrapper properties generated by the compilers). For that property wrappers should implemented coding protocols. Additionally it's possible to extendion standard library decoding containers with decoding methods for specific types of wrappers without writing decoding code for the whole type - container will pick up extension methods if the types in the signature match. This way we can customise decoding for non-standard cases in various ways.
+
+Simplest example of a property wrapper would be if we make a `ISO8601Date` a property wrapper. This only requires to add a `@propertyWrapper` annotation to this type. It already implements decoding for dates so we don't need to change that. What changes is only how we declare properties of this type.
+
+Before:
+
+```swift
+struct Appointment: Decodable {
+  let time: Date
+	
+  init(from decoder: Decoder) throws {
+    let values = try decoder.container(keyedBy: CodingKeys.self)
+    time = try values.decode(ISO8601Date.self, forKey: .time)
+  }
+}
+```
+
+After:
+
+```swift
+struct Appointment: Decodable {
+  @ISO8601Date
+  let time: Date
+}
+```
+
+Reference: https://speakerdeck.com/alisoftware/and-thats-a-wrap?slide=25
+
+Another example is pruned/strict arrays decoding. Pruned strategy means that any array element that fails decoding will be discarded when with strict strategy the whole decoding will fail on any failure in decoding of any array element (default behaviour). To implement prune strategy we use `FailableDecodable` that decodes its value with `value = try? container.decode(T.self)` or `decodeArraySafely` method. With property wrapper we wouldn't need to to call this method manually or change the type of the property. We can also implement more strategies, like non-empty, using the same property wrapper.
+
+Before:
+
+```swift
+struct DemographicsListDTO: Decodable {
+  let demographics: [DemographicDTO]
+	
+  init(from decoder: Decoder) throws {
+    let values = try decoder.container(keyedBy: CodingKeys.self)
+    let demographics = try values.decodeArraySafely([DemographicDTO].self, forKey: .demographics)
+  }
+}
+```
+
+After:
+
+```swift
+struct DemographicsListDTO: Decodable {
+  @ArrayDecodable<Prune, DemographicDTO>
+  let demographics: [DemographicDTO]
+}
+```
+
+Providing defaults for absent values has a bit tricky ergonomics, as this approach does not allow to pass arbitrary parameters to property wrapper initialisers when they are being decoded (so `@Default(value: true)` is possible programmatically but won't affect decoding). So instead of using values we use types which would correspond to common defaults, i.e. `True`, `False`, `Empty` etc:
```

Comment:
```suggestion
Providing defaults for absent values has slightly tricky ergonomics, as this approach does not allow to pass arbitrary parameters to property wrapper initialisers when they are being decoded (so `@Default(value: true)` is possible programmatically but won't affect decoding). So instead of using values we use types which would correspond to common defaults, i.e. `True`, `False`, `Empty` etc:
```